linopy/constraints.py

```diff
@@ -14,58 +14,37 @@
 import numpy as np
 import pandas as pd
 import xarray as xr
+from deprecation import deprecated
 from numpy import arange, array
 from scipy.sparse import coo_matrix
 from xarray import DataArray, Dataset
 
 from linopy import expressions, variables
 from linopy.common import (
     _merge_inplace,
-    has_assigned_model,
     has_optimized_model,
     is_constant,
     replace_by_map,
 )
 
 
-class Constraint(DataArray):
+@dataclass(repr=False)
+class Constraint:
     """
-    Constraint container for storing constraint labels.
+    Projection to a single constraint in a model.
 
     The Constraint class is a subclass of xr.DataArray hence most xarray
     functions can be applied to it.
     """
 
-    __slots__ = ("_cache", "_coords", "_indexes", "_name", "_variable", "model")
-
-    def __init__(self, *args, **kwargs):
-
-        # workaround until https://github.com/pydata/xarray/pull/5984 is merged
-        if isinstance(args[0], DataArray):
```

Comment:
This seems to suggest that the class supports initialization from a `DataArray` object passed as first positional parameter `args[0]`. The new interface is not backwards compatible in this way. But I found only [one place](https://github.com/PyPSA/linopy/blob/composition-design/linopy/constraints.py#L253) where the constructor `Constraint()` is used directly, so I guess this should be fine. I don't really understand the comment and how it is related to the linked pull request.

Reply:
The PR 5984 was merged for 0.20.2 , ie below what we currently support as oldest possible version so not relevant anymore.

Indeed, I changed the `Constraint` semantics to only ever possibly refer to a full named `Constraint` that is already represented within a `Constraints` container (i considered renaming it to `BoundConstraint` to make this more obvious).

The `AnonymousConstraint` seemed to fulfill the need for anything that is not attached to a model yet.

Reply:
Note also that this means that you cannot use `Constraint` to represent a slice of a constraint, ie. there is no way to `sel` into a constraint and still use a `Constraint` object to hold it. (i only understood that feature through converting `Variable` where this is needed).

Reply:
Hmm the workaround for PR 5984 is obsolete, but I assume there was a reason why `linopy.Constraint` had to support initialization from a `DataArray` object, right? I can't see such a place changed in this PR here nor could I find such a place in the `linopy` repository, so I'd guess: this is either outdated or used in some downstream repository (PyPSA?).